### PR TITLE
[WIP] Refactor CMake build system.

### PR DESCRIPTION
**Pull request by geektoni**
This PR contains major refactors of the BioDynaMo's build system. It is still WIP and some features are still missing.

Here a list of some features which were implemented and some of the missing ones:
- [x] Do not require to source `biodynamo-env.sh` for building biodynamo;
- [x] Make install step optional (same file structure inside build dir as if it would be installed);
- [x] ParaView and ROOT are installed automatically if they are not found;
- [x] Automatic generation of scripts which install missing prerequisites (optional and required);
- [x] Add missing dependency detection (valgrind, mkdocs, lcov, ...) Don't rely on the fact that these are included in $PATH
- [x] Avoid code duplication between `UseBioDynaMo.cmake` and `CMakeLists.txt`

### CODE REVIEW

File: prerequisites.sh [hunk 1]
```diff
@@ -14,26 +14,68 @@
...
 Description:
   This script installs the prerequisites of BioDynaMo, but not BioDynaMo
-  itself. Script install.sh installs both prerequisites and BioDynaMo.
-No Arguments
+  itself.
+Usage:
+  prerequisites.sh <os_name> <type_of_prerequisites>
+
+  <os_name>                 It is the name of your current operating system. The ones
+                            supported by BioDynaMo are:
```
Comment (breitwieserCern) [inline prerequisites.sh]: Use function to detect supported OS

File: prerequisites.sh [hunk 2]
```diff
@@ -14,26 +14,68 @@
 # -----------------------------------------------------------------------------
 
 # This script installs the prerequisites of BioDynaMo, but not BioDynaMo
-# itself. Script install.sh installs both prerequisites and BioDynaMo.
+# itself.
 # Arguments:
-#  $1 path to the biodynamo project directory
+#  $1 operative systems
```
Comment (breitwieserCern) [inline prerequisites.sh]: operating system

File: doc/user_guide/docs/prerequisites.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+  * **cmake**: Set of tools for automate building, testing of software (for */usr/bin/cmake*);
+  * **make**: Build automation tool (for */usr/bin/make*);
+  * **gcc**: GNU C compiler (for */usr/bin/gcc*);
+  * **g++**: GNU C++ compiler (for */usr/bin/g++*);
+  * **libopenmpi-dev**: Development files for OpenMPI (Open Source Message Passing Interface);
+  * **libomp-dev**: Development files for OpenMP (API for multiprocessor programming);
+  * **libnuma-dev**: Development files for NUMA (simple programming interface to the policy supported by the Linux kernel);
+  * **libtbb-dev**: Development files for TBB (C++ template library developed by Intel for parallel programming);
+  * **freeglut3-dev**: Development files for GLUT (OpenGL Utility Toolkit);
```
Comment (breitwieserCern) [inline doc/user_guide/docs/prerequisites.md]: I think this is only needed by ParaView -> optional
Also change it in the cmake files

**Comment by breitwieserCern** [comment]
Remove folder `util/installation/travis-linux`

File: install.sh [hunk 4]
```diff
@@ -28,5 +28,8 @@ BDM_PROJECT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
 # include util functions
 . $BDM_PROJECT_DIR/util/installation/common/util.sh
 
+# Install all prerequisites
+$BDM_PROJECT_DIR/prerequisites.sh $(DetectOs) all
+
 # call install script for the detected OS
 CallOSSpecificScript $BDM_PROJECT_DIR install.sh
```
Comment (breitwieserCern) [inline install.sh]: Let's call `util/installation/common/install.sh` directly.
All os specific scripts forward the call 
Remove all os specific `install.sh` files

File: install.sh [hunk 5]
```diff
@@ -28,5 +28,8 @@ BDM_PROJECT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
 # include util functions
 . $BDM_PROJECT_DIR/util/installation/common/util.sh
 
+# Install all prerequisites
+$BDM_PROJECT_DIR/prerequisites.sh $(DetectOs) all
```
Comment (breitwieserCern) [inline install.sh]: Let's only detect the OS if the user didn't specify one.
Similar to the CMake solution

File: util/installation/common/install.sh [hunk 6]
```diff
@@ -34,18 +34,48 @@ BDM_PROJECT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)/../../.."
...
+  fi
+
+  . /etc/profile.d/modules.sh
+  module load mpi
+fi
+set -e
 
 # perform a clean release build
 BUILD_DIR=$BDM_PROJECT_DIR/build
 CleanBuild $BUILD_DIR
```
Comment (breitwieserCern) [inline util/installation/common/install.sh]: Let's forward BDM_OS to CMake to avoid inconsistencies

**Comment by geektoni** [comment]
> Remove folder `util/installation/travis-linux`

All the comments were addressed.

File: doc/user_guide/docs/installation.md [hunk 7]
```diff
@@ -12,10 +12,13 @@ cd biodynamo
 !!! important
 
     1. After the installation you need to restart your terminal.
-      In every new terminal execute `source <path-to-bdm-installation>/biodynamo-env.sh`
+      In every new terminal execute `source <path-to-bdm-installation>/biodynamo/bin/thisbdm.sh`
```
Comment (breitwieserCern) [inline doc/user_guide/docs/installation.md]: `source <path-to-bdm-installation>/bin/thisbdm.sh`

File: doc/user_guide/docs/installation.md [hunk 8]
```diff
@@ -12,10 +12,13 @@ cd biodynamo
...
       to use BioDynaMo!
-
-    2. BioDynaMo uses a customized version of ParaView.
+      
+    2. It is also possible to use the library without running `make install`. You will just need to source `thisbdm.sh` from the build
+       directory: `source <path-to-bdm-build-dir>/biodynamo/bin/thisbdm.sh`
```
Comment (breitwieserCern) [inline doc/user_guide/docs/installation.md]: `source <path-to-bdm-build-dir>/install/bin/thisbdm.sh`

File: doc/user_guide/docs/prerequisites.md [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+### Installation
+```bash
+sudo apt-get install -y cmake make gcc g++ \
+libopenmpi-dev libomp-dev libnuma-dev libtbb-dev \
+libpthread-stubs0-dev \
+python3 python3-pip
+```
+
+```bash
```
Comment (breitwieserCern) [inline doc/user_guide/docs/prerequisites.md]: Let's mark them as optional
Also for other the other systems

File: install.sh [hunk 10]
```diff
@@ -13,11 +13,14 @@
...
-if [[ $# -ne 0 ]]; then
+if [ $# -ge 2 ]; then
   echo "ERROR: Wrong number of arguments.
+Usage:
+  install.sh [<os>]
 Description:
   This script installs/updates the currently checked out version of biodynamo
-No Arguments"
+Arguments:
+  <os>  Specify manually which operative system is being used."
```
Comment (breitwieserCern) [inline install.sh]: ... which operating system ...

File: util/housekeeping/run-clang-format.sh [hunk 11]
```diff
@@ -44,7 +44,9 @@ MODE=$1
 shift
 
 if [ "$#" == "0" ]; then
-  echo "Warning: No files to process"
+  echo "Warning: No files to process."
+  echo -e "This may happen because the repository has no origin/master branch set."
+  echo "Try to add it using 'git remote add' or run 'make format-all' instead."
```
Comment (breitwieserCern) [inline util/housekeeping/run-clang-format.sh]: Move this check and the error message to `CppStyleGuideChecks.cmake`

File: util/housekeeping/run-clang-tidy.sh [hunk 12]
```diff
@@ -54,8 +54,10 @@ MODE=$1
 shift
 
 if [ "$#" == "0" ]; then
-  echo "Warning: No files to process"
-  exit 0;
+    echo "Warning: No files to process."
+    echo -e "This may happen because the repository has no origin/master branch set."
+    echo "Try to add it using 'git remote add' or run 'make tidy-all' instead."
+    exit 0;
```
Comment (breitwieserCern) [inline util/housekeeping/run-clang-tidy.sh]: same as `run-clang-format.sh`

File: util/installation/common/util.sh [hunk 13]
```diff
@@ -265,10 +236,10 @@ function EchoFinishThisStep {
   fi
 
   EchoInfo "To complete this step execute:"
-  EchoInfo "    ${BDM_ECHO_BOLD}${BDM_ECHO_UNDERLINE}source $1/biodynamo-env.sh"
+  EchoInfo "    ${BDM_ECHO_BOLD}${BDM_ECHO_UNDERLINE}source $1/biodynamo/bin/thisbdm.sh"
```
Comment (breitwieserCern) [inline util/installation/common/util.sh]: `source $1/bin/thisbdm.sh`

File: util/installation/common/util.sh [hunk 14]
```diff
@@ -265,10 +236,10 @@ function EchoFinishThisStep {
...
   EchoInfo "This command must be executed in every terminal before you build or use BioDynaMo."
   EchoInfo "To avoid this additional step add it to your $(BashrcFile) file:"
-  EchoInfo "    echo \"source $1/biodynamo-env.sh\" >> $(BashrcFile)"
+  EchoInfo "    echo \"source $1/biodynamo/bin/thisbdm.sh\" >> $(BashrcFile)"
```
Comment (breitwieserCern) [inline util/installation/common/util.sh]: `source $1/bin/thisbdm.sh`

File: util/installation/ubuntu-16.04/prerequisites.sh [hunk 15]
```diff
@@ -13,122 +13,28 @@
...
-  rm $THIRD_PARTY_DIR/qt/lib/cmake/Qt5Gui/Qt5Gui_QGtk3ThemePlugin.cmake
-  touch $THIRD_PARTY_DIR/qt/lib/cmake/Qt5Gui/Qt5Gui_QGtk3ThemePlugin.cmake
-
-  EchoSuccess "Installation of prerequisites finished successfully!"
-  EchoFinishThisStep $BDM_INSTALL_DIR
-  echo ""
-}
-
-Install
+# Install required packages
```
Comment (breitwieserCern) [inline util/installation/ubuntu-16.04/prerequisites.sh]: add `sudo apt update`

File: .travis.yml [hunk 16]
```diff
@@ -11,7 +11,7 @@ matrix:
       env: SCRIPT="util/run-inside-docker.sh ubuntu-16.04 util/travis-ci/default-build.sh"
     # Default build on osx
     - os: osx
-      compiler: clang
+      language: generic
```
Comment (breitwieserCern) [inline .travis.yml]: why was this change required?

File: doc/dev_guide/docs/build.md [hunk 17]
```diff
@@ -5,20 +5,82 @@ To build BioDynaMo from source execute the following commands:
...
+make
 
-mkdir build && cd build
-cmake .. && make -j4
+# (Optional) Installs the library 
 make install
 ```
 
+The script `prerequisites.sh` is used to install all the dependencies needed by BioDynaMo. You will need
+to run it before actually calling `cmake` and `make`. It will also choose the specific dependencies given the operative systems.
```
Comment (breitwieserCern) [inline doc/dev_guide/docs/build.md]: operating systems

File: CMakeLists.txt [hunk 18]
```diff
@@ -1,83 +1,191 @@
...
+# This is necessary as we for example call `clang -v` and parse its output. But on a localized
+# program, the output parsing is much more error prone as certrain strings we're looking for
+# could be missing or be in a different order. To prevent those errors, let's just force all
+# output to use the default C locale which is more or less identical on all systems.
+set(ENV{LANG} C)
+
+# Project name and current version. We do not specify any language for
+# this project. Therefore, we will be able to check manually which compiler
+# we detected on the system.
+project(BioDynaMo LANGUAGES VERSION "0.1.0")
```
Comment (breitwieserCern) [inline CMakeLists.txt]: remove version string

File: CMakeLists.txt [hunk 19]
```diff
@@ -1,83 +1,191 @@
...
+include(CheckCompiler)
+
+# Directory in which BioDynaMo will be installed. It can be provided
+# by the user itself.
+if (NOT DEFINED ${BDM_INSTALL_DIR})
+    set(BDM_INSTALL_DIR $ENV{HOME}/.bdm)
+endif()
+
+# Specify some directories which will be used to structure BioDynaMo build.
+set(CMAKE_BIODYNAMO_ROOT        "${CMAKE_BINARY_DIR}")
```
Comment (breitwieserCern) [inline CMakeLists.txt]: `CMAKE_BUILD_ROOT` would be more intuitive

File: CMakeLists.txt [hunk 20]
```diff
@@ -1,83 +1,191 @@
...
-  set(OpenMP_C_FLAGS -fopenmp=libomp)
-  set(OpenMP_CXX_FLAGS -fopenmp=libomp)
-  include_directories(/usr/local/Cellar/biodynamo)
+# Check if we have a compatible openmp compiler
+find_package(OpenMP)
+if (OpenMP_FOUND)
+    link_directories(${OpenMP_CXX_LIBRARIES})
 else()
-  find_package(OpenMP REQUIRED)
+    MESSAGE(FATAL_ERROR "We did not found any support for OpenMP. Please run ./prerequisites.sh again before
```
Comment (breitwieserCern) [inline CMakeLists.txt]: We did not find OpenMP on your system ...

File: CMakeLists.txt [hunk 21]
```diff
@@ -1,83 +1,191 @@
...
+calling cmake. If this does not work, then you current compiler might not support OpenMP (this happens for instance
```
Comment (breitwieserCern) [inline CMakeLists.txt]: ... then your current ...

File: CMakeLists.txt [hunk 22]
```diff
@@ -1,83 +1,191 @@
...
   # must be before include(${ROOT_USE_FILE}) otherwise test compile fails
-  find_package(Threads REQUIRED)
+  find_package(Threads)
 endif()
 
 if(cuda)
   find_package(CUDA)
+  if (NOT CUDA_FOUND)
+    MESSAGE("CUDA was not found in your system, therefore it was disabled to let you compile BioDynaMo. In order to
+enable this functionality please install CUDA for your system. This can be done automatically with the prerequistes.sh
```
Comment (breitwieserCern) [inline CMakeLists.txt]: CUDA is not installed in `prerequesites.sh`
Point user to cuda website

File: CMakeLists.txt [hunk 23]
```diff
@@ -1,83 +1,191 @@
...
+    SET(cuda OFF)
+  endif()
 endif()
+
 if(opencl)
   find_package(OpenCL)
+  if (NOT OpenCL_FOUND)
+      MESSAGE("OpenCL was not found in your system, therefore it was disabled to let you compile BioDynaMo. In order to
+enable this functionality please install OpenCL for your system. This can be done automatically with the prerequistes.sh
+script offered by the library. Then run again cmake (with the -Dopencl=ON flag).")
```
Comment (breitwieserCern) [inline CMakeLists.txt]: OpenCL is not installed in `prerequesites.sh`
Point user to cuda website

File: CMakeLists.txt [hunk 24]
```diff
@@ -98,37 +206,88 @@ if (NUMA_FOUND)
...
-  find_package(ParaView 5.3 QUIET OPTIONAL_COMPONENTS vtkPVPythonCatalyst
-    vtkPVVTKExtensionsDefault vtkPVServerManagerRendering)
-  if(${ParaView_FOUND})
-    message(STATUS "Found ParaView")
+
+    # If we specify a custom location for ParaView then we need to specify also a custom location
+    # for Qt.
+    if ((DEFINED ENV{ParaView_DIR} AND NOT DEFINED ENV{Qt5_DIR}) OR (NOT DEFINED ENV{ParaView_DIR} AND DEFINED ENV{Qt5_DIR}))
+        MESSAGE(FATAL_ERROR "\nFATAL ERROR \nYou need to specify both ParaView_DIR and Qt5_DIR in order to build\
+ BioDynaMo with those library. This is needed only if you want to use your own versions of ParaView and Qt.\
```
Comment (breitwieserCern) [inline CMakeLists.txt]: in order to build BioDynaMo.

File: CMakeLists.txt [hunk 25]
```diff
@@ -98,37 +206,88 @@ if (NUMA_FOUND)
...
+if (vtune)
+    find_package(VTune)
+    if(${VTune_FOUND})
+        include_directories(${VTUNE_INCLUDE_DIR})
+        add_definitions("-DUSE_VTUNE")
+        set(BDM_REQUIRED_LIBRARIES ${BDM_REQUIRED_LIBRARIES} ${VTUNE_LIBRARIES})
+    else()
+        MESSAGE("VTune was not found in your system, therefore it was disabled. You will not be able to run the perfomance
+checks. In order to enable this functionality please install VTune for your system. This can be done automatically with
+the prerequistes.sh script offered by the library. Then run again cmake (with the -Dvtune=ON flag).")
```
Comment (breitwieserCern) [inline CMakeLists.txt]: vtune is not installed with prerequisites

**Comment by breitwieserCern** [comment]
If git is not found: which version string is used?
Let's rename: `runBioDynamoTestMain` to `unit-tests`

File: .travis.yml (hunk 16 above)
Comment (geektoni) [reply .travis.yml]: This is needed to prevent Travis to set `CXX` and `CC` environmental variables. The automatic detection of the compiler works only if those two are not set.

File: CMakeLists.txt (hunk 19 above)
Comment (geektoni) [reply CMakeLists.txt]: I have opted for `CMAKE_BIODYNAMO_BUILD_ROOT`. `CMAKE_BUILD_ROOT` seems to be related to the ROOT framework.

File: cmake/CppStyleGuideChecks.cmake [hunk 26]
```diff
@@ -82,9 +80,25 @@ if (GIT_FOUND)
...
+  # - make format
+  # - make tidy
+  # - make check-cpplint-all
+  if ("${HAVE_ORIGIN_REPOSITORY}" EQUAL "1")
+    add_clang_format_target(format "${CHANGED_SRC_FILES_ORIGIN_MASTER}" )
+    add_clang_tidy_target(tidy "${CHANGED_SRC_FILES_ORIGIN_MASTER}" )
+    add_cpplint_target(check-cpplint "${CHANGED_SRC_FILES_ORIGIN_MASTER}" )
+  else()
+    MESSAGE(WARNING "\nWe did not detected any remote origin/master branch. Therefore targets \
+like 'make format', 'make tidy' and 'make check-cpplint-all' will not be available. Use 'make format-all', 'make tidy-all' and 'make check-cpplint-staged' instead. \
```
Comment (breitwieserCern) [inline cmake/CppStyleGuideChecks.cmake]: make check-cpplint   (without -all)

File: CMakeLists.txt [hunk 27]
```diff
@@ -1,83 +1,191 @@
...
+Then run again cmake (with the -Dcuda=ON flag).")
+    SET(cuda OFF)
+  endif()
 endif()
+
 if(opencl)
   find_package(OpenCL)
+  if (NOT OpenCL_FOUND)
+      MESSAGE("OpenCL was not found in your system, therefore it was disabled to let you compile BioDynaMo. In order to
+enable this functionality please install OpenCL for your system. Please have a look at https://developer.nvidia.com/opencl.
```
Comment (breitwieserCern) [inline CMakeLists.txt]: Please have a look at the website of your graphic card vendor on how to install OpenCL.
